src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -983,12 +1155,11 @@ - (void)_changeInternalState:(MTRInternalDeviceState)state
 
         /* BEGIN DRAGONS: This is a huge hack for a specific use case, do not rename, remove or modify behavior here */
         // TODO: This should only be called for thread devices
-        id<MTRDeviceDelegate> delegate = _weakDelegate.strongObject;
-        if ([delegate respondsToSelector:@selector(_deviceInternalStateChanged:)]) {
-            dispatch_async(_delegateQueue, ^{
-                [(id) delegate _deviceInternalStateChanged:self];
-            });
-        }
+        [self _callDelegatesWithBlock:^(id delegate) {
+            if ([delegate respondsToSelector:@selector(_deviceInternalStateChanged:)]) {
```

Comment:
I think it would be good to perform all the `respondsToSelector` calls when we add the delegate and cache this as a bit mask on the DelegateInfo, so we can avoid async-dispatching blocks that end up doing nothing.